File: tools/commands/ci/github/release-tag-rollback.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../../rust/common.rs"]
mod common;

use std::{env, process::ExitCode};

#[derive(Debug)]
struct Context {
    repository: String,
    sha: String,
    tag: String,
    token: String,
}

#[derive(Clone, Debug)]
struct RemoteTag {
    commit_sha: String,
    object_sha: String,
}

fn main() -> ExitCode {
    match rollback_unpublished_tag() {
        Ok(result) => {
            println!("{result}");
            ExitCode::SUCCESS
        }
        Err(error) => {
            eprintln!("{error}");
            ExitCode::from(1)
        }
    }
}

fn rollback_unpublished_tag() -> Result<String, String> {
    let context = assert_rollback_context()?;
    let tag_ref = format!("refs/tags/{}", context.tag);
    let remote = remote_tag_state(
        &git(&[
            "ls-remote",
            "--tags",
            "origin",
            &tag_ref,
            &format!("{tag_ref}^{{}}"),
        ])?
        .stdout,
        &context.tag,
    )?;
    let Some(remote) = remote else {
        return Ok(format!("Release tag {} was already absent.", context.tag));
    };
    if remote.commit_sha != context.sha {
        return Err(format!(
            "Refusing to delete {}: remote tag resolves to {}, not event SHA {}.",
            context.tag, remote.commit_sha, context.sha
        ));
    }

    assert_no_github_release(&context)?;
    git(&[
        "fetch",
        "--quiet",
        "--force",
        "origin",
        &format!("{tag_ref}:{tag_ref}"),
    ])?;
    let local_object_sha = git(&["rev-parse", &tag_ref])?.stdout.trim().to_string();
    let local_commit_sha = git(&["rev-parse", &format!("{tag_ref}^{{}}")])?
        .stdout
        .trim()
        .to_string();
    if local_object_sha != remote.object_sha || local_commit_sha != context.sha {
        return Err(format!(
            "Refusing to delete {}: fetched tag identity does not match the audited remote tag.",
            context.tag
        ));
    }

    assert_no_github_release(&context)?;
    git(&[
        "push",
        &format!("--force-with-lease={tag_ref}:{}", remote.object_sha),
        "origin",
        &format!(":{tag_ref}"),
    ])?;
    Ok(format!(
        "Rolled back unpublished release tag {}.",
        context.tag
    ))
}

fn assert_rollback_context() -> Result<Context, String> {
    let tag = env::var("GITHUB_REF_NAME").unwrap_or_default();
    let sha = env::var("GITHUB_SHA").unwrap_or_default();
    let repository = env::var("GITHUB_REPOSITORY").unwrap_or_default();
    let preflight_result = env::var("RELEASE_PREFLIGHT_RESULT").unwrap_or_default();
    let token = env::var("GITHUB_TOKEN").unwrap_or_default();

    if env::var("GITHUB_REF_TYPE").unwrap_or_default() != "tag" {
        return Err(format!(
            "Release rollback requires a tag event, got {}.",
            env::var("GITHUB_REF_TYPE").unwrap_or_else(|_| "unknown".to_string())
        ));
    }
    parse_release_version(&tag)?;
    if !is_strict_v_tag(&tag) {
        return Err(format!(
            "Release rollback requires a v-prefixed release tag, got {tag}."
        ));
    }
    if !is_full_sha(&sha) {
        return Err(format!(
            "Release rollback requires a full event SHA, got {sha}."
        ));
    }
    if !is_owner_repo(&repository) {
        return Err(format!(
            "Release rollback requires an owner/repository identity, got {repository}."
        ));
    }
    if !matches!(preflight_result.as_str(), "failure" | "cancelled") {
        return Err(format!(
            "Refusing to roll back {tag}: release preflight concluded {}.",
            if preflight_result.is_empty() {
                "unknown"
            } else {
                &preflight_result
            }
        ));
    }
    if token.is_empty() {
        return Err(
            "Release rollback requires GITHUB_TOKEN to check for an existing release.".to_string(),
        );
    }
    Ok(Context {
        repository,
        sha,
        tag,
        token,
    })
}

fn remote_tag_state(output: &str, tag: &str) -> Result<Option<RemoteTag>, String> {
    let tag_ref = format!("refs/tags/{tag}");
    let peeled_ref = format!("{tag_ref}^{{}}");
    let mut object_sha = None;
    let mut commit_sha = None;
    for line in output.lines().filter(|line| !line.trim().is_empty()) {
        let parts = line.split_whitespace().collect::<Vec<_>>();
        let [sha, reference] = parts.as_slice() else {
            continue;
        };
        if *reference == tag_ref {
            object_sha = Some((*sha).to_string());
        } else if *reference == peeled_ref {
            commit_sha = Some((*sha).to_string());
        }
    }
    let Some(object_sha) = object_sha else {
        return Ok(None);
    };
    let commit_sha = commit_sha.unwrap_or_else(|| object_sha.clone());
    if !is_full_sha(&object_sha) || !is_full_sha(&commit_sha) {
        return Err(format!("Remote tag {tag} returned malformed object IDs."));
    }
    Ok(Some(RemoteTag {
        commit_sha,
        object_sha,
    }))
}

fn assert_no_github_release(context: &Context) -> Result<(), String> {
    let api_url =
        env::var("GITHUB_API_URL").unwrap_or_else(|_| "https://api.github.com".to_string());
    let url = format!(
        "{}/repos/{}/releases/tags/{}",
        api_url.trim_end_matches('/'),
        context.repository,
        context.tag
    );
    let output = common::run_capture(
        "curl",
        &[
            "-sS",
            "-L",
            "-w",
            "\n%{http_code}",
            "-H",
            "Accept: application/vnd.github+json",
            "-H",
            &format!("Authorization: Bearer {}", context.token),
            "-H",
            "X-GitHub-Api-Version: 2022-11-28",
            &url,
        ],
    )?;
    let (body, status) = output
        .stdout
        .rsplit_once('\n')
        .ok_or_else(|| "curl did not return an HTTP status".to_string())?;
    match status.trim() {
        "404" => Ok(()),
        "200" => Err(format!(
            "Refusing to delete {}: a GitHub Release already exists.",
            context.tag
        )),
        other => Err(format!(
            "Could not prove that {} is unpublished: GitHub Releases API returned {}{}.",
            context.tag,
            other,
            if body.trim().is_empty() {
                String::new()
            } else {
                format!(": {}", body.trim())
            }
        )),
    }
}

fn git(args: &[&str]) -> Result<common::CommandOutput, String> {
    common::run_capture("git", args)
}

fn parse_release_version(ref_name: &str) -> Result<(), String> {
    let name = ref_name.strip_prefix('v').unwrap_or(ref_name);
    let stable = name.split_once('-').map(|(value, _)| value).unwrap_or(name);
    let parts = stable.split('.').collect::<Vec<_>>();
    let [major, minor, patch] = parts.as_slice() else {
        return Err(format!(
            "Release tag must look like vMAJOR.MINOR.PATCH[-PRERELEASE], got {ref_name}"
        ));
    };
    major
        .parse::<u64>()
        .map_err(|_| invalid_release(ref_name))?;
    minor
        .parse::<u64>()
        .map_err(|_| invalid_release(ref_name))?;
    patch
        .parse::<u64>()
        .map_err(|_| invalid_release(ref_name))?;
    Ok(())
}

fn invalid_release(ref_name: &str) -> String {
    format!("Release tag must look like vMAJOR.MINOR.PATCH[-PRERELEASE], got {ref_name}")
}

fn is_strict_v_tag(value: &str) -> bool {
    value.strip_prefix('v').is_some_and(|rest| {
        let stable = rest.split_once('-').map(|(value, _)| value).unwrap_or(rest);
        let parts = stable.split('.').collect::<Vec<_>>();
        matches!(parts.as_slice(), [major, minor, patch] if is_u64(major) && is_u64(minor) && is_u64(patch))
    })
}

fn is_u64(value: &str) -> bool {
    !value.is_empty() && value.bytes().all(|byte| byte.is_ascii_digit())
}

fn is_full_sha(value: &str) -> bool {
    value.len() == 40 && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

fn is_owner_repo(value: &str) -> bool {
    let parts = value.split('/').collect::<Vec<_>>();
    matches!(parts.as_slice(), [owner, repo] if is_repo_part(owner) && is_repo_part(repo))
}

fn is_repo_part(value: &str) -> bool {
    !value.is_empty()
        && value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'.' | b'-'))
}

File: tools/support/compat/github/release-tag-rollback.mjs
import { spawnSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { parseReleaseVersion } from "./release-platforms.mjs";

const DEFAULT_GIT_TIMEOUT_MS = 30_000;
const rollbackResults = new Set(["failure", "cancelled"]);

function runGit(args, cwd = process.cwd(), timeoutMs = DEFAULT_GIT_TIMEOUT_MS) {
  const result = spawnSync("git", args, {
    cwd,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    timeout: timeoutMs,
  });
  if (result.error?.code === "ETIMEDOUT") {
    throw new Error(`git ${args.join(" ")} timed out after ${timeoutMs}ms.`);
  }
  if (result.error != null) throw result.error;
  if (result.signal != null) {
    throw new Error(`git ${args.join(" ")} was terminated by ${result.signal}.`);
  }
  if (result.status !== 0) {
    const detail = [result.stdout, result.stderr].filter(Boolean).join("\n").trim();
    throw new Error(
      [`git ${args.join(" ")} failed with exit ${result.status}`, detail]
        .filter(Boolean)
        .join("\n"),
    );
  }
  return result;
}

export function assertRollbackContext(env) {
  const tag = env.GITHUB_REF_NAME ?? "";
  const sha = env.GITHUB_SHA ?? "";
  const repository = env.GITHUB_REPOSITORY ?? "";
  const preflightResult = env.RELEASE_PREFLIGHT_RESULT ?? "";
  const token = env.GITHUB_TOKEN ?? "";

  if (env.GITHUB_REF_TYPE !== "tag") {
    throw new Error(
      `Release rollback requires a tag event, got ${env.GITHUB_REF_TYPE ?? "unknown"}.`,
    );
  }
  parseReleaseVersion(tag);
  if (!/^v\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?$/.test(tag)) {
    throw new Error(`Release rollback requires a v-prefixed release tag, got ${tag}.`);
  }
  if (!/^[0-9a-f]{40}$/.test(sha)) {
    throw new Error(`Release rollback requires a full event SHA, got ${sha}.`);
  }
  if (!/^[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+$/.test(repository)) {
    throw new Error(`Release rollback requires an owner/repository identity, got ${repository}.`);
  }
  if (!rollbackResults.has(preflightResult)) {
    throw new Error(
      `Refusing to roll back ${tag}: release preflight concluded ${preflightResult || "unknown"}.`,
    );
  }
  if (token === "") {
    throw new Error("Release rollback requires GITHUB_TOKEN to check for an existing release.");
  }
  return { preflightResult, repository, sha, tag, token };
}

export function remoteTagState(output, tag) {
  const refs = new Map(
    output
      .trim()
      .split(/\r?\n/)
      .filter(Boolean)
      .map((line) => line.trim().split(/\s+/, 2).reverse()),
  );
  const objectSha = refs.get(`refs/tags/${tag}`);
  if (objectSha == null) return undefined;
  const commitSha = refs.get(`refs/tags/${tag}^{}`) ?? objectSha;
  if (!/^[0-9a-f]{40}$/.test(objectSha) || !/^[0-9a-f]{40}$/.test(commitSha)) {
    throw new Error(`Remote tag ${tag} returned malformed object IDs.`);
  }
  return { commitSha, objectSha };
}

async function assertNoGithubRelease({ apiUrl, fetchImpl, repository, tag, token }) {
  const response = await fetchImpl(
    `${apiUrl}/repos/${repository}/releases/tags/${encodeURIComponent(tag)}`,
    {
      headers: {
        Accept: "application/vnd.github+json",
        Authorization: `Bearer ${token}`,
        "X-GitHub-Api-Version": "2022-11-28",
      },
    },
  );
  if (response.status === 404) return;
  if (response.status === 200) {
    throw new Error(`Refusing to delete ${tag}: a GitHub Release already exists.`);
  }
  const detail = (await response.text()).trim();
  throw new Error(
    `Could not prove that ${tag} is unpublished: GitHub Releases API returned ${response.status}${detail ? `: ${detail}` : ""}.`,
  );
}

export async function rollbackUnpublishedTag({
  env = process.env,
  cwd = process.cwd(),
  fetchImpl = globalThis.fetch,
  git = (args) => runGit(args, cwd),
} = {}) {
  const context = assertRollbackContext(env);
  const tagRef = `refs/tags/${context.tag}`;
  const readRemote = () =>
    remoteTagState(
      git(["ls-remote", "--tags", "origin", tagRef, `${tagRef}^{}`]).stdout,
      context.tag,
    );
  const remote = readRemote();
  if (remote == null) {
    return { deleted: false, reason: "already-absent", tag: context.tag };
  }
  if (remote.commitSha !== context.sha) {
    throw new Error(
      `Refusing to delete ${context.tag}: remote tag resolves to ${remote.commitSha}, not event SHA ${context.sha}.`,
    );
  }

  await assertNoGithubRelease({
    apiUrl: (env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, ""),
    fetchImpl,
    ...context,
  });

  git(["fetch", "--quiet", "--force", "origin", `${tagRef}:${tagRef}`]);
  const localObjectSha = git(["rev-parse", tagRef]).stdout.trim();
  const localCommitSha = git(["rev-parse", `${tagRef}^{}`]).stdout.trim();
  if (localObjectSha !== remote.objectSha || localCommitSha !== context.sha) {
    throw new Error(
      `Refusing to delete ${context.tag}: fetched tag identity does not match the audited remote tag.`,
    );
  }

  // Recheck after fetching so a Release created during the tag audit cannot
  // slip through immediately before the destructive operation.
  await assertNoGithubRelease({
    apiUrl: (env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, ""),
    fetchImpl,
    ...context,
  });

  git(["push", `--force-with-lease=${tagRef}:${remote.objectSha}`, "origin", `:${tagRef}`]);
  return { deleted: true, tag: context.tag };
}

const entrypoint = process.argv[1]
  ? fileURLToPath(import.meta.url) === path.resolve(process.argv[1])
  : false;
if (entrypoint) {
  rollbackUnpublishedTag()
    .then((result) => {
      console.log(
        result.deleted
          ? `Rolled back unpublished release tag ${result.tag}.`
          : `Release tag ${result.tag} was already absent.`,
      );
    })
    .catch((error) => {
      console.error(error instanceof Error ? error.message : String(error));
      process.exit(1);
    });
}
